File: stockbot/src/services/scheduler.py
from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.cron import CronTrigger
import logging
import asyncio
from src.config import get_settings
from src.services.redis_service import db
from src.services.stock_checker import checker
from src.services.notifier import notifier

settings = get_settings()
logger = logging.getLogger(__name__)

class SchedulerService:
    def __init__(self):
        self.scheduler = AsyncIOScheduler()
        self.is_running_check = False

    def start(self):
        # Schedule check based on config interval
        trigger = CronTrigger(minute=f"*/{settings.CHECK_INTERVAL_MINUTES}")
        self.scheduler.add_job(self.run_check, trigger)
        self.scheduler.start()
        logger.info(f"⏰ Scheduler started (every {settings.CHECK_INTERVAL_MINUTES} minutes)")
        
        # Initial check after 10s
        asyncio.create_task(self._delayed_initial_check())

    async def _delayed_initial_check(self):
        await asyncio.sleep(10)
        await self.run_check()

    def stop(self):
        self.scheduler.shutdown()
        logger.info("⏹️ Scheduler stopped")

    async def run_check(self):
        if self.is_running_check:
            logger.warning("⚠️ Check already running, skipping...")
            return

        self.is_running_check = True
        logger.info("🔄 Running scheduled stock check...")

        try:
            products = await db.get_all_products()
            
            if not products:
                logger.info("📭 No products to check")
                self.is_running_check = False
                return

            logger.info(f"📦 Checking {len(products)} product(s)...")

            for product in products:
                await self._check_product(product)
                await asyncio.sleep(2)  # Crawl delay

            logger.info("✅ Stock check completed")

        except Exception as e:
            logger.error(f"❌ Error during stock check: {e}")
        finally:
            self.is_running_check = False

    async def _check_product(self, product: dict):
        try:
            url = product['url']
            old_status = product.get('status')
            
            result = await checker.check_stock(url)
            
            if result['status'] == 'error':
                logger.warning(f"⚠️ Error checking {product.get('name')}: {result.get('error')}")
                return

            new_status = result['status']
            
            # Update Redis
            await db.update_product_status(url, new_status, {
                "name": result.get('name') or product.get('name'),
                "imageUrl": result.get('imageUrl') or product.get('imageUrl')
            })

            # Notify if changed
            if old_status != new_status and old_status != 'unknown':
                logger.info(f"📢 Status change: {product.get('name')} - {old_status} -> {new_status}")
                
                subscribers = await db.get_subscribers(url)
                if subscribers:
                    updated_product = await db.get_product(url)
                    await notifier.notify_users(list(subscribers), updated_product, old_status, new_status)

        except Exception as e:
            logger.error(f"❌ Error checking product {product.get('url')}: {e}")

    async def force_check(self):
        logger.info("🔄 Forcing immediate stock check...")
        await self.run_check()

scheduler = SchedulerService()

File: stockbot/src/services/scheduler.js
import cron from 'node-cron';
import config from '../config/index.js';
import redis from './redis.js';
import stockChecker from './stockChecker.js';
import notifier from './notifier.js';

class Scheduler {
    constructor() {
        this.job = null;
        this.isRunning = false;
    }

    /**
     * Start the scheduler
     */
    start() {
        const intervalMinutes = config.checker.intervalMinutes;
        const cronExpression = `*/${intervalMinutes} * * * *`;

        console.log(`⏰ Starting scheduler (every ${intervalMinutes} minutes)`);

        this.job = cron.schedule(cronExpression, async () => {
            await this.runCheck();
        });

        // Run initial check after 10 seconds
        setTimeout(() => this.runCheck(), 10000);
    }

    /**
     * Stop the scheduler
     */
    stop() {
        if (this.job) {
            this.job.stop();
            this.job = null;
            console.log('⏹️ Scheduler stopped');
        }
    }

    /**
     * Run a stock check for all tracked products
     */
    async runCheck() {
        if (this.isRunning) {
            console.log('⚠️ Check already running, skipping...');
            return;
        }

        this.isRunning = true;
        console.log('🔄 Running scheduled stock check...');

        try {
            const products = await redis.getAllProducts();

            if (products.length === 0) {
                console.log('📭 No products to check');
                this.isRunning = false;
                return;
            }

            console.log(`📦 Checking ${products.length} product(s)...`);

            for (const product of products) {
                await this.checkProduct(product);
                // Small delay between checks
                await new Promise(resolve => setTimeout(resolve, 2000));
            }

            console.log('✅ Stock check completed');

        } catch (error) {
            console.error('❌ Error during stock check:', error.message);
        } finally {
            this.isRunning = false;
        }
    }

    /**
     * Check a single product and notify on status change
     */
    async checkProduct(product) {
        try {
            const oldStatus = product.status;
            const result = await stockChecker.checkStock(product.url);

            if (result.status === 'error') {
                console.log(`⚠️ Error checking ${product.name}: ${result.error}`);
                return;
            }

            const newStatus = result.status;

            // Update product in Redis
            await redis.updateProductStatus(product.url, newStatus, {
                name: result.name || product.name,
                imageUrl: result.imageUrl || product.imageUrl,
            });

            // Check if status changed
            if (oldStatus !== newStatus && oldStatus !== 'unknown') {
                console.log(`📢 Status change: ${product.name} - ${oldStatus} → ${newStatus}`);

                // Get subscribers and notify them
                const subscribers = await redis.getSubscribers(product.url);

                if (subscribers.length > 0) {
                    const updatedProduct = await redis.getProduct(product.url);
                    await notifier.notifyUsers(
                        subscribers,
                        updatedProduct,
                        oldStatus,
                        newStatus
                    );
                }
            }

        } catch (error) {
            console.error(`❌ Error checking product ${product.url}:`, error.message);
        }
    }

    /**
     * Force an immediate check for all products
     */
    async forceCheck() {
        console.log('🔄 Forcing immediate stock check...');
        await this.runCheck();
    }
}

// Singleton instance
export const scheduler = new Scheduler();
export default scheduler;
